enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,130 @@
+
+#### Safeguaring control-plane components
+
+Gatekeeper uses a single admission webhook to for constraints enforcement. We assume the same solution will be used for the mutation feature. The control-plane resources can be excluded from gatekeeper interference by specifying an appropriate namespace selector on the webhook.
+Gatekeeper would be installed/configured by an operator, which would adjust the webhook namespace selector (based on some GatekeeperInstallConfig CRD).
+Also the default gatekeeper webhook configuration is using "failurePolicy​: ​Ignore", we assume this would be extended to the mutation feature (more on this: https://github.com/open-policy-agent/gatekeeper#admission-webhook-fail-open-status). Any resources which would slip into the system unmodified would later be corrected using the additional relabeling feature (implemented in the second phase).
+
+Currently we can only specify the namespaces to be included. The operator would have to update the webhook config every time a new namespace is added. Webhook specific exludes in kubernetes are discussed [here|https://github.com/kubernetes/kubernetes/issues/92157]
+
+#### Performance
+
+Multi-pod deployments of Gatekeeper is already implemented.
+The relabeling feature would have to be implemented as daemonset to avoid performance issues.
+
+#### Risk assesment
+
+There are multiple risks involved with delivering the feature using gatekeeper:
+
+- ETA: the time frame for the first part of the feature (mutating webhook) is quite tight, even though there are currently 3 people (different projects) commited to work on this feature
+- We have no maintainers on the project, which could be a bottleneck with getting the feature code in, or getting some required features in.
+- The requirements and design of the second part of the feature (drift correction) is nowhere near ready. There are still open questions on how do resolve issues like constraint conflicts or convergence.
+- The upstream code might not be of the required quality by the required time.
+
+### Other projects
+
+Kyverno (https://github.com/nirmata/kyverno) was also looked at but was mostly rejected. It does not provide all the needed functionality, the community is mostly backed by one company and seems less lively than gatekeeper and it seems to be in bad technical shape (getting it to run was quite problematic as oposed to gatekeeper).
+
+## Graduation Criteria
+
+#### Dev Preview -> Tech Preview
```

Comment:
I miss a detailed discussion of the API versions. The new API will be alpha. Will we be able to guarantee forward compatibility, i.e. will customer objects created in alpha survive a version update when the API is promoted to beta? Usually alpha APIs don't have that property (at least in upstream Kubernetes).

Reply:
I hope the API currently discussed will be the final one, or it will differ in details only. At least this is the impression the gatekeeper community discussions gives.
The original idea of the rego based API was discarded in favor of the simple API, so I assume that is out of the question.
The current proposal can be found here: https://docs.google.com/document/d/1MdchNFz9guycX__QMGxpJviPaT_MZs8iXaAFqCvoXYQ/edit?ts=5f73fb77#

Reply:
which means it is declared v1 from the start?

I think you have to add "v1beta1" API level as a requirement for leaving dev preview phase.

Reply:
Added